### PR TITLE
adds write_cal to hera_cal.io

### PR DESCRIPTION
**Pull request by nkern**
`hera_cal.io.write_cal` is a method to format gain dictionaries into `pyuvdata.UVCal` structures, and to then write them out as `calfits` files.

### CODE REVIEW

File: hera_cal/io.py [hunk 1]
```diff
@@ -243,15 +247,131 @@ def load_cal(input_cal, return_meta=False):
...
+    '''Format gain solution dictionary into pyuvdata.UVCal and write to file
+
+    Arguments:
+        fname : type=str, output file basename
+        gains : type=dictionary, holds complex gain solutions. keys are antenna + pol
+                tuple pairs, e.g. (2, 'x'), and keys are 2D complex ndarrays with time
+                along [0] axis and freq along [1] axis.
+        freqs : type=ndarray, holds unique frequencies channels in Hz
+        times : type=ndarray, holds unique times of integration centers in Julian Date
+        pols : type=ndarray, holds unique polarization of gains in 'x' and/or 'y' form
```
Comment (zakiali) [inline hera_cal/io.py]: "...gains in string form. Can be either 'x' or 'y'".

File: hera_cal/io.py [hunk 2]
```diff
@@ -243,15 +247,131 @@ def load_cal(input_cal, return_meta=False):
...
+    freq_array = freq_array[None, :]
+    spw_array = np.arange(Nspws)
+    channel_width = np.median(np.diff(freq_array))
+
+    # form gain, flags and qualities
+    gain_array = np.empty((Nants_data, Nspws, Nfreqs, Ntimes, Njones), np.complex)
+    flag_array = np.empty((Nants_data, Nspws, Nfreqs, Ntimes, Njones), np.bool)
+    quality_array = np.empty((Nants_data, Nspws, Nfreqs, Ntimes, Njones), np.float)
+    for i, p in enumerate(pol_array):
+        for j, a in enumerate(antenna_numbers):
```
Comment (zakiali) [inline hera_cal/io.py]: This should be a loop over ant_array. From pyuvdata docs, ant array is the index of the antenna in data arrays.

Comment (nkern) [reply hera_cal/io.py]: but its an enumerate, so `j` is the index, which is being used to index into `gain_array`. this must be `antenna_numbers` because `a` is used to index into the `gains` dictionary. right?

File: hera_cal/tests/test_io.py [hunk 3]
```diff
@@ -226,12 +226,39 @@ def test_load_cal(self):
...
+            for j, a in enumerate(ants):
+                gains[(a, p)] = np.ones((Ntimes, Nfreqs), np.complex)
+                quality[(a, p)] = np.ones((Ntimes, Nfreqs), np.float) * 2
+                flags[(a, p)] = np.zeros((Ntimes, Nfreqs), np.bool)
+
+        # test basic execution
+        uvc = io.write_cal("ex.calfits", gains, freqs, times, pols, flags=flags, quality=quality,
+                           overwrite=True, return_uvc=True, write_file=True)
+        self.assertTrue(os.path.exists("ex.calfits"))
+        self.assertAlmostEqual(uvc.gain_array[0,0,0,0,0], (1+0j))
```
Comment (zakiali) [inline hera_cal/tests/test_io.py]: Do you want to loop over all antennas and pols and do these assertions?
I understand it's redundant, but there may be some weird edge cases not found by this test.

File: hera_cal/io.py (hunk 2 above)
Comment (zakiali) [reply hera_cal/io.py]: In this case it can be either antenna_numbers or ant_array, but from the docs, ant_array is the parameter you want to use to do this, not antenna_numbers. This is more for consistency than anything. My understanding is that antenna_numbers can be in any order and is just a way to label/name the antennas where ant_array is the array you look at for indexing into the data arrays. It's the same loop you have just replacing antenna_numbers with ant_array.

Comment (zakiali) [reply hera_cal/io.py]: oh wait, just realized that ant_array is just an array from 0, nants-1. Why?

Comment (zakiali) [reply hera_cal/io.py]: Oh wait, I misunderstood the docs. I would be happy if it was something like for j, a in zip(ant_array, antenna_numbers)

Comment (nkern) [reply hera_cal/io.py]: but I think that is exactly the same as what I have here: `enumerate(antenna_numbers)`

File: hera_cal/io.py (hunk 1 above)
Comment (nkern) [reply hera_cal/io.py]: got rid of pol argument actually, can get it from the gains dictionary

File: hera_cal/tests/test_io.py (hunk 3 above)
Comment (nkern) [reply hera_cal/tests/test_io.py]: meh, not a fan of the idea of looping through the whole array... instead of looping over I just checked the sum of the entire array is what I would expect

**Comment by nkern** [comment]
thanks for the comments!

**Comment by nkern** [comment]
addresses #178
